  <div id="fileHeader">
    <h1>failing_mock_argument_constraints_spec.rb</h1>
    <table class="header-table">
    <tr class="top-aligned-row">
      <td><strong>Path:</strong></td>
      <td>spec/spec/mocks/failing_mock_argument_constraints_spec.rb</td>
    </tr>
    <tr class="top-aligned-row">
      <td><strong>Last Update:</strong></td>
      <td>Mon Dec 29 13:11:11 +0100 2008</td>
    </tr>
    </table>
  </div>
 <!-- banner header -->

  <div id="bodyContent" >
    <h2>Source Code</h2>
    <pre>require File.dirname(__FILE__) + '/../../spec_helper.rb'

module Spec
  module Mocks
    describe &quot;failing MockArgumentConstraints&quot; do
      before(:each) do
        @mock = mock(&quot;test mock&quot;)
        @reporter = Mock.new(&quot;reporter&quot;, :null_object =&gt; true)
      end
      
      after(:each) do
        @mock.rspec_reset
      end

      it &quot;should reject non boolean&quot; do
        @mock.should_receive(:random_call).with(boolean())
        lambda do
          @mock.random_call(&quot;false&quot;)
        end.should raise_error(MockExpectationError)
      end

      it &quot;should reject non numeric&quot; do
        @mock.should_receive(:random_call).with(an_instance_of(Numeric))
        lambda do
          @mock.random_call(&quot;1&quot;)
        end.should raise_error(MockExpectationError)
      end
      
      it &quot;should reject non string&quot; do
        @mock.should_receive(:random_call).with(an_instance_of(String))
        lambda do
          @mock.random_call(123)
        end.should raise_error(MockExpectationError)
      end
      
      it &quot;should reject goose when expecting a duck&quot; do
        @mock.should_receive(:random_call).with(duck_type(:abs, :div))
        lambda { @mock.random_call(&quot;I don't respond to :abs or :div&quot;) }.should raise_error(MockExpectationError)
      end

      it &quot;should fail if regexp does not match submitted string&quot; do
        @mock.should_receive(:random_call).with(/bcd/)
        lambda { @mock.random_call(&quot;abc&quot;) }.should raise_error(MockExpectationError)
      end
      
      it &quot;should fail if regexp does not match submitted regexp&quot; do
        @mock.should_receive(:random_call).with(/bcd/)
        lambda { @mock.random_call(/bcde/) }.should raise_error(MockExpectationError)
      end
      
      it &quot;should fail for a hash w/ wrong values&quot; do
        @mock.should_receive(:random_call).with(:a =&gt; &quot;b&quot;, :c =&gt; &quot;d&quot;)
        lambda do
          @mock.random_call(:a =&gt; &quot;b&quot;, :c =&gt; &quot;e&quot;)
        end.should raise_error(MockExpectationError, /Mock 'test mock' expected :random_call with \(\{(:a=&gt;\&quot;b\&quot;, :c=&gt;\&quot;d\&quot;|:c=&gt;\&quot;d\&quot;, :a=&gt;\&quot;b\&quot;)\}\) but received it with \(\{(:a=&gt;\&quot;b\&quot;, :c=&gt;\&quot;e\&quot;|:c=&gt;\&quot;e\&quot;, :a=&gt;\&quot;b\&quot;)\}\)/)
      end
      
      it &quot;should fail for a hash w/ wrong keys&quot; do
        @mock.should_receive(:random_call).with(:a =&gt; &quot;b&quot;, :c =&gt; &quot;d&quot;)
        lambda do
          @mock.random_call(&quot;a&quot; =&gt; &quot;b&quot;, &quot;c&quot; =&gt; &quot;d&quot;)
        end.should raise_error(MockExpectationError, /Mock 'test mock' expected :random_call with \(\{(:a=&gt;\&quot;b\&quot;, :c=&gt;\&quot;d\&quot;|:c=&gt;\&quot;d\&quot;, :a=&gt;\&quot;b\&quot;)\}\) but received it with \(\{(\&quot;a\&quot;=&gt;\&quot;b\&quot;, \&quot;c\&quot;=&gt;\&quot;d\&quot;|\&quot;c\&quot;=&gt;\&quot;d\&quot;, \&quot;a\&quot;=&gt;\&quot;b\&quot;)\}\)/)
      end
      
      it &quot;should match against a Matcher&quot; do
        lambda do
          @mock.should_receive(:msg).with(equal(3))
          @mock.msg(37)
        end.should raise_error(MockExpectationError, &quot;Mock 'test mock' expected :msg with (equal 3) but received it with (37)&quot;)
      end
      
      it &quot;should fail no_args with one arg&quot; do
        lambda do
          @mock.should_receive(:msg).with(no_args)
          @mock.msg(37)
        end.should raise_error(MockExpectationError, &quot;Mock 'test mock' expected :msg with (no args) but received it with (37)&quot;)
      end
      
      it &quot;should fail hash_including with missing key&quot; do
         lambda do
           @mock.should_receive(:msg).with(hash_including(:a =&gt; 1))
           @mock.msg({})
         end.should raise_error(MockExpectationError, &quot;Mock 'test mock' expected :msg with (hash_including(:a=&gt;1)) but received it with ({})&quot;)
      end

      it &quot;should fail with block constraints&quot; do
        lambda do
          @mock.should_receive(:msg).with {|arg| arg.should == :received }
          @mock.msg :no_msg_for_you
        end.should raise_error(Spec::Expectations::ExpectationNotMetError, /expected: :received.*\s*.*got: :no_msg_for_you/)
      end
            
    end
  end
end
</pre>
    </div>